### PR TITLE
feat(compiler): Add extended diagnostic to warn when missing let on n…

### LINKED ISSUE #14641: fix: better error messages for NgFor
**I'm submitting a ...**  (check one with "x")

[x] bug report => search github for a similar issue or PR before submitting
[ ] feature request
[ ] support request

**Current behavior**
While there have been a lot of improvements in the error messages space in Angular, it turns out that there's one case left that can causes many hours of debugging, because the error message is confusing.

When forgetting the `let` keyword in `NgFor`'s micro syntax, that error message points the developer in a wrong direction. **Basically, doing this**:

```
<ANY *ngFor="item of items"></ANY>
```

Throws this error:

```
Can't bind to 'ngFor' since it isn't a known property of 'ANY'.
1. If 'ANY' is an Angular component and it has 'ngFor' input, then verify that it is part of this module.
2. If 'ANY' is a Web Component then add "CUSTOM_ELEMENTS_SCHEMA" to the '@NgModule.schemas' of this component to suppress this message.
```

This error messages usually occurs when one forgets to add/export a directive/component (in this case `NgFor`), but the actually issue is that the developer forgot the `let` keyword in the micro syntax.

Unfortunately this caused me once 1.5 hours of debugging and wondering why Angular thinks I haven't declared `NgFor` anywhere, while it is in fact properly setup.

I'm seeing more and more people running into this, so I thought it's worth bringing up.

**Expected behavior**
It would be cool if we can throw a more specific error message, or maybe extend the existing one to tell the developer that she needs to double-check the micro syntax.

**Plunker for reproduction

https://plnkr.co/edit/SAI2cYUjuiVXcg4sL8vz?p=preview (check the console)

* **Angular version:** 2.x.

### PR DESCRIPTION
**Pull request by jessicajaniuk**
…gForOf

In the case that a user accidentally forgot the let keyword, they dont get a very clear indicator of there being a problem.
They get an issue in the template iteration at runtime. This diagnostic will warn the user when the let keyword is missing.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [x] Other... Please describe: Adding an extended template diagnostic

## What is the current behavior?

The user gets no warning about missing the `let` keyword in ngFor and instead gets an unhelpful error at runtime.

Issue Number: resolves #14641

## What is the new behavior?
The template diagnostic will warn the user ahead of time.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: packages/compiler-cli/src/ngtsc/typecheck/extended/checks/missing_ngforof_let/index.ts [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+class MissingNgForOfLetCheck extends TemplateCheckWithVisitor<ErrorCode.MISSING_NGFOROF_LET> {
+  override code = ErrorCode.MISSING_NGFOROF_LET as const;
+
+  override visitNode(
+      ctx: TemplateContext<ErrorCode.MISSING_NGFOROF_LET>, component: ts.ClassDeclaration,
+      node: TmplAstNode|AST): NgTemplateDiagnostic<ErrorCode.MISSING_NGFOROF_LET>[] {
+    const isTemplate = node instanceof TmplAstTemplate;
+    if (!(isTemplate) || (isTemplate && !(node.templateAttrs.length))) return [];
+    const attr = node.templateAttrs.find(x => x.name === 'ngForOf');
+    if (attr == null) return [];
```
Comment (alxhub) [inline packages/compiler-cli/src/ngtsc/typecheck/extended/checks/missing_ngforof_let/index.ts:29]:
```suggestion
    if (attr === undefined) {
      return [];
    }
```

File: packages/compiler-cli/src/ngtsc/typecheck/extended/checks/missing_ngforof_let/index.ts [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+    if (node.variables.length > 0) return [];
```
Comment (alxhub) [inline packages/compiler-cli/src/ngtsc/typecheck/extended/checks/missing_ngforof_let/index.ts:31]: Style (in the compiler at least) is to always use blocks/braces with conditionals/loops/etc.

```suggestion
    if (node.variables.length > 0) {
      return [];
    }
```

File: packages/compiler-cli/src/ngtsc/typecheck/extended/checks/missing_ngforof_let/index.ts [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    const errorString = 'Your ngForOf is missing a value. Did you forget to add the `let` keyword?';
+    const diagnostic = ctx.makeTemplateDiagnostic(node.sourceSpan, errorString);
```
Comment (alxhub) [inline packages/compiler-cli/src/ngtsc/typecheck/extended/checks/missing_ngforof_let/index.ts:33]: We should consider whether there's a better node to highlight for the error. `node` here is the entire `TmplAstTemplate` - the whole `*ngFor` element including its body. That's a lot of red squiggly lines in the IDE.

Instead, we could highlight _just_ the `*ngFor` attribute name itself. `attr.keySpan` I believe would do that.

**Review by alxhub — APPROVED** [review]

File: packages/compiler-cli/src/ngtsc/typecheck/extended/checks/missing_ngforof_let/index.ts (hunk 3 above)
Comment (jessicajaniuk) [reply packages/compiler-cli/src/ngtsc/typecheck/extended/checks/missing_ngforof_let/index.ts]: Looks like `attr.keySpan` only refers to the `of` keyword. `attr.sourceSpan` gives us `of items` in the test. It seems to skip the variable `item` in the string, but I think it's probably a better highlight than the whole template. I've switched it to that.

File: packages/compiler-cli/src/ngtsc/typecheck/extended/checks/missing_ngforof_let/index.ts [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+ * Will return diagnostic information when `let` is missing.
+ */
+class MissingNgForOfLetCheck extends TemplateCheckWithVisitor<ErrorCode.MISSING_NGFOROF_LET> {
+  override code = ErrorCode.MISSING_NGFOROF_LET as const;
+
+  override visitNode(
+      ctx: TemplateContext<ErrorCode.MISSING_NGFOROF_LET>, component: ts.ClassDeclaration,
+      node: TmplAstNode|AST): NgTemplateDiagnostic<ErrorCode.MISSING_NGFOROF_LET>[] {
+    const isTemplate = node instanceof TmplAstTemplate;
+    if (!(isTemplate) || (isTemplate && !(node.templateAttrs.length))) return [];
```
[the remark this answers is not in the document]
Comment (jessicajaniuk) [reply packages/compiler-cli/src/ngtsc/typecheck/extended/checks/missing_ngforof_let/index.ts]: Thanks. I've updated this.

File: packages/compiler-cli/src/ngtsc/typecheck/extended/checks/missing_ngforof_let/index.ts (hunk 1 above)
Comment (jessicajaniuk) [reply packages/compiler-cli/src/ngtsc/typecheck/extended/checks/missing_ngforof_let/index.ts]: Fixed.

File: packages/compiler-cli/src/ngtsc/typecheck/extended/checks/missing_ngforof_let/index.ts (hunk 2 above)
Comment (jessicajaniuk) [reply packages/compiler-cli/src/ngtsc/typecheck/extended/checks/missing_ngforof_let/index.ts]: Fixed.

File: packages/compiler-cli/src/ngtsc/typecheck/extended/test/checks/missing_ngforof_let/missing_ngforof_let_spec.ts [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+            'export class TestCmp { items: [] = [{\'name\': \'diana\'}, {\'name\': \'prince\'}] }'
+      }]);
+      const sf = getSourceFileOrError(program, fileName);
+      const component = getClass(sf, 'TestCmp');
+      const extendedTemplateChecker = new ExtendedTemplateCheckerImpl(
+          templateTypeChecker, program.getTypeChecker(), [missingNgForOfLet], {} /* options */);
+      const diags = extendedTemplateChecker.getDiagnosticsForComponent(component);
+      expect(diags.length).toBe(0);
+    });
+  });
```
Comment (atscott) [inline packages/compiler-cli/src/ngtsc/typecheck/extended/test/checks/missing_ngforof_let/missing_ngforof_let_spec.ts:64]: Could you add a test to ensure the diagnostic is not produced when using the valid longhand version:

```
<ng-template ngFor let-item [ngForOf]="items">
  {{item}}
</ng-template>
```

Comment (jessicajaniuk) [reply packages/compiler-cli/src/ngtsc/typecheck/extended/test/checks/missing_ngforof_let/missing_ngforof_let_spec.ts:82]: Sounds good. I added it.

**Review by atscott — APPROVED** [review]
reviewed-for: public-api

**Review by alxhub — APPROVED** [review]
Reviewed-for: public-api

**Pull request closed without merging**